src/services/ConfigService.ts

```diff
@@ -0,0 +1,77 @@
+import * as path from "path";
+import * as fs from "fs";
+import NeedleConfig from "../models/NeedleConfig";
+import DO_NOT_TOUCH_defaultConfig from "../config.json";
+import Setting from "../models/enums/Setting";
+
+export default class ConfigService {
+	private readonly directoryPath: string;
+	private readonly cache = new Map<string, NeedleConfig>();
+
+	constructor(directoryPath: string) {
+		this.directoryPath = path.resolve(__dirname, "../../", process.env.CONFIGS_PATH || directoryPath);
+	}
+
+	public get(guildId: string): NeedleConfig {
+		const guildConfig = this.cache.get(guildId) ?? this.readFromFile(guildId);
+		const defaultConfigCopy = this.getDefault();
+
+		if (guildConfig) {
+			guildConfig.settings = Object.assign({}, defaultConfigCopy.settings, guildConfig?.settings);
+		}
+
+		return Object.assign({}, defaultConfigCopy, guildConfig);
+	}
+
+	public set(guildId: string, config: NeedleConfig) {
+		if (!fs.existsSync(this.directoryPath)) {
+			fs.mkdirSync(this.directoryPath);
+		}
+
+		config.threadChannels = config.threadChannels?.filter(val => val !== null && val !== undefined);
+
+		// Only save messages that are different from the defaults
+		const defaultConfigCopy = this.getDefault();
+		if (defaultConfigCopy.settings && config.settings) {
+			for (const [key, message] of Object.entries(config.settings)) {
+				if (message !== defaultConfigCopy.settings[key as keyof typeof Setting]) continue;
+				delete config.settings[key as keyof typeof Setting];
+			}
+		}
+
+		const configPath = this.getPath(guildId);
+		fs.writeFileSync(configPath, JSON.stringify(config), { encoding: "utf-8" });
+		this.cache.set(guildId, config);
+	}
+
+	public delete(guildId: string): boolean {
+		const configPath = this.getPath(guildId);
+		if (!fs.existsSync(configPath)) return false;
+
+		fs.rmSync(configPath);
+		this.cache.delete(guildId);
+
+		console.log(`Deleted data for guild ${guildId}`);
+		return true;
+	}
+
+	public getDefault(): NeedleConfig {
+		// Return a clone so we don't mess with the actual default config (touch the variable here and only here)
+		return JSON.parse(JSON.stringify(DO_NOT_TOUCH_defaultConfig));
```

Comment:
Not seeing any immediate performance benefits? Tested on https://www.measurethat.net/Benchmarks/Show/18532/0/jsonstringify-jsonparse-vs-structuredclone

Reply:
![image](https://user-images.githubusercontent.com/35617441/187310827-e0aaf2df-2ead-44cd-af64-1cfe80f10828.png)

Reply:
I assumed that just cloning an object instead of coverting to JSON then back to a JS object would be quicker; you can resolve this then.

Reply:
Since Node LTS is on version 16 I'll stick to that to make self-hosting more accessible with Needle. I know some people are stuck with 16 that self-host Needle currently, and I don't want to prevent them from self-hosting.